File: src/gigaphone/engine/plan.py
"""`gigaphone plan` — boundaries → plan records, plus the unresolved[] list (DESIGN §5, §11).

A descriptor that resolved to no boundary is *unresolved* (the ambiguous ~20% the
deterministic pass can't localize) and is surfaced via the resolution protocol — never
silently skipped (golden principle 8).
"""

from __future__ import annotations

from dataclasses import dataclass

from gigaphone.core.boundary import BoundaryKind
from gigaphone.core.model import Boundary, Descriptor
from gigaphone.core.plan_record import PlanRecord


@dataclass
class Unresolved:
    descriptor_id: str
    match_call: str
    question: str


@dataclass
class Plan:
    records: list[PlanRecord]
    unresolved: list[Unresolved]

    @property
    def fixable(self) -> list[PlanRecord]:
        return [r for r in self.records if r.failure_modes]


def build_plan(descriptors: list[Descriptor], boundaries: list[Boundary]) -> Plan:
    records = [
        PlanRecord(
            boundary=f"{b.path}:{b.range.line}",
            language="python",
            provider_or_framework=b.provider_or_framework,
            kind=b.kind,
            tools_covered=list(b.tools_covered),
            failure_modes=list(b.failure_modes),
            complete_output_fields=list(b.complete_output_fields),
            source=b.source,
        )
        for b in boundaries
    ]
    resolved_calls = {b.call for b in boundaries}
    unresolved = [
        Unresolved(d.id, d.match_call, _question_for(d))
        for d in descriptors
        if d.match_call not in resolved_calls and d.kind != BoundaryKind.LLM
    ]
    return Plan(records=records, unresolved=unresolved)


def _question_for(d: Descriptor) -> str:
    if d.kind == BoundaryKind.AGENT_CALL:
        return (
            f"Could not localize `{d.match_call}` (agent_call). Which function dispatches "
            "the sub-agent and returns its result? (The sub-agent itself is a black box — we "
            "trace only this boundary.)"
        )
    return (
        f"Could not localize `{d.match_call}` ({d.kind.value}). "
        "Which function consumes its result and returns it to the agent loop?"
    )

File: src/engine/plan.ts
/**
 * `gigaphone plan` — boundaries → plan records, plus the unresolved[] list (DESIGN §5, §11).
 *
 * A descriptor that resolved to no boundary is *unresolved* (the ambiguous ~20% the
 * deterministic pass can't localize) and is surfaced via the resolution protocol — never
 * silently skipped (golden principle 8).
 */

import { BoundaryKind } from "../core/boundary.js";
import type { Boundary, Descriptor } from "../core/model.js";
import { PlanRecord } from "../core/planRecord.js";

export interface Unresolved {
  descriptorId: string;
  matchCall: string;
  question: string;
}

export class Plan {
  constructor(
    readonly records: PlanRecord[],
    readonly unresolved: Unresolved[],
  ) {}

  get fixable(): PlanRecord[] {
    return this.records.filter((r) => r.failureModes.length > 0);
  }
}

export function buildPlan(descriptors: Descriptor[], boundaries: Boundary[]): Plan {
  const records = boundaries.map(
    (b) =>
      new PlanRecord({
        boundary: `${b.path}:${b.range.line}`,
        language: "python",
        providerOrFramework: b.providerOrFramework,
        kind: b.kind,
        toolsCovered: [...b.toolsCovered],
        failureModes: [...b.failureModes],
        completeOutputFields: [...b.completeOutputFields],
        source: b.source,
      }),
  );
  const resolvedCalls = new Set(boundaries.map((b) => b.call));
  const unresolved: Unresolved[] = descriptors
    .filter((d) => !resolvedCalls.has(d.matchCall))
    .map((d) => ({ descriptorId: d.id, matchCall: d.matchCall, question: questionFor(d) }));
  return new Plan(records, unresolved);
}

function questionFor(d: Descriptor): string {
  if (d.kind === BoundaryKind.AGENT_CALL) {
    return (
      `Could not localize \`${d.matchCall}\` (agent_call). Which function dispatches ` +
      "the sub-agent and returns its result? (The sub-agent itself is a black box — we " +
      "trace only this boundary.)"
    );
  }
  return (
    `Could not localize \`${d.matchCall}\` (${d.kind}). ` +
    "Which function consumes its result and returns it to the agent loop?"
  );
}
